File: samples/fizzbuzz.cpp
#include <iostream>
#include <vector>
#include <functional>
#include <algorithm>
#include <string>
#include <numeric>

std::string fizzBuzzOperation(int num) {
    std::string result = "";
    if (num % 3 == 0) result += "Fizz";
    if (num % 5 == 0) result += "Buzz";
    return result.empty() ? std::to_string(num) : result;
}

int main() {
    std::vector<int> numbers(15);
    std::iota(numbers.begin(), numbers.end(), 1);
    
    std::function<std::string(int)> operation = fizzBuzzOperation;
    
    std::for_each(numbers.begin(), numbers.end(), [&operation](int num) {
        std::cout << operation(num) << std::endl;
    });

    return 0;
}

File: samples/fizzbuzz.rs
trait Operation {
    fn perform(&self, num: i32) -> Result<String, &'static str>;
}

struct FizzBuzz;

impl Operation for FizzBuzz {
    fn perform(&self, num: i32) -> Result<String, &'static str> {
        match (num % 3, num % 5) {
            (0, 0) => Ok("FizzBuzz".to_string()),
            (0, _) => Ok("Fizz".to_string()),
            (_, 0) => Ok("Buzz".to_string()),
            _ => Ok(num.to_string()),
        }
    }
}

fn process_numbers(numbers: Vec<i32>, op: &dyn Operation) -> Vec<Result<String, &'static str>> {
    numbers.iter().map(|&num| op.perform(num)).collect()
}

fn main() {
    let numbers = (1..=15).collect();
    let fizzbuzz = FizzBuzz;
    let results = process_numbers(numbers, &fizzbuzz);
    for result in results {
        println!("{}", result.unwrap_or_else(|e| e.to_string()));
    }
}
